src/cpp/session/modules/SessionCopilot.cpp

```
@@ -0,0 +1,894 @@
+   //
+   // TODO: Does this need to be a state machine? Is it possible that we
+   // receive some incomplete requests here?
+   std::size_t startIndex = 0;
+
+   while (startIndex < stdOut.length())
+   {
+      // Find the split.
+      auto splitIndex = stdOut.find("\r\n\r\n", startIndex);
+      if (splitIndex == std::string::npos)
+      {
+         ELOG("Internal error: parsing response failed.");
+         break;
+      }
+
+      // Extract the header text.
+      std::string headerText = string_utils::substring(stdOut, startIndex, splitIndex + 4);
+
+      // Parse the headers.
+      core::http::Headers headers;
+      std::istringstream iss(headerText);
+      core::http::parseHeaders(iss, &headers);
+
+      // Check for content length.
+      std::string contentLength = core::http::headerValue(headers, "Content-Length");
+      if (contentLength.empty())
+      {
+         ELOG("Internal error: response contains no Content-Length header.");
+         break;
+      }
+
+      // Consume that text.
+      auto bodyStart = splitIndex + 4;
+      auto bodyEnd = bodyStart + safe_convert::stringTo<int>(contentLength, 0);
+      std::string bodyText = string_utils::substring(stdOut, bodyStart, bodyEnd);
+      s_pendingResponses.push(bodyText);
+
+      if (copilotLogLevel() > 2)
+      {
+         std::cerr << std::endl;
+         std::cerr << "RESPONSE" << std::endl;
+         std::cerr << "------------------" << std::endl;
+         std::cerr << bodyText << std::endl;
+         std::cerr << "------------------" << std::endl;
+         std::cerr << std::endl << std::endl;
+      }
+
+      // Update the start index.
+      startIndex = bodyEnd;
+   }
+}
+
+void onStderr(ProcessOperations& operations, const std::string& stdErr)
+{
+   std::cerr << stdErr << std::endl;
+}
+
+void onExit(int status)
+{
+   s_agentPid = -1;
+
+   if (status != 0)
+   {
+      ELOG("Agent exited with status {}.", status);
```

We should suppress error logging if the exit happens after a shutdown has been requested.